## HW #1
Student name: Olha Tanyuk

1.	Install Python (if you don’t have it already), and install NLTK.  

2.	Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

3.	Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

4.	Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.  

5.	Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.


## Answers:
### 1. Python was installed.
### 2. Getting Started with NLTK:

In [1]:
import nltk
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

In [2]:
from nltk.book import *

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [3]:
text1

<Text: Moby Dick by Herman Melville 1851>

In [4]:
text2

<Text: Sense and Sensibility by Jane Austen 1811>

### Searching Text

Here we look up the word 'monstrous' in Moby Dick

In [5]:
text1.concordance("monstrous")

Displaying 11 of 11 matches:
ong the former , one was of a most monstrous size . ... This came towards us , 
ON OF THE PSALMS . " Touching that monstrous bulk of the whale or ork we have r
ll over with a heathenish array of monstrous clubs and spears . Some were thick
d as you gazed , and wondered what monstrous cannibal and savage could ever hav
that has survived the flood ; most monstrous and most mountainous ! That Himmal
they might scout at Moby Dick as a monstrous fable , or still worse and more de
th of Radney .'" CHAPTER 55 Of the Monstrous Pictures of Whales . I shall ere l
ing Scenes . In connexion with the monstrous pictures of whales , I am strongly
ere to enter upon those still more monstrous stories of them which are to be fo
ght have been rummaged out of this monstrous cabinet there is no telling . But 
of Whale - Bones ; for Whales of a monstrous size are oftentimes cast up dead u


Here we search Sense and Sensibility for the word affection

In [6]:
text2.concordance("affection")

Displaying 25 of 79 matches:
, however , and , as a mark of his affection for the three girls , he left them
t . It was very well known that no affection was ever supposed to exist between
deration of politeness or maternal affection on the side of the former , the tw
d the suspicion -- the hope of his affection for me may warrant , without impru
hich forbade the indulgence of his affection . She knew that his mother neither
rd she gave one with still greater affection . Though her late conversation wit
 can never hope to feel or inspire affection again , and if her home be uncomfo
m of the sense , elegance , mutual affection , and domestic comfort of the fami
, and which recommended him to her affection beyond every thing else . His soci
ween the parties might forward the affection of Mr . Willoughby , an equally st
 the most pointed assurance of her affection . Elinor could not be surprised at
he natural consequence of a strong affection in a young and ardent mind . This 
 opinion . 

Here we search the book of Genesis to find out how long some people lived

In [7]:
text3.concordance("lived")

Displaying 25 of 38 matches:
ay when they were created . And Adam lived an hundred and thirty years , and be
ughters : And all the days that Adam lived were nine hundred and thirty yea and
nd thirty yea and he died . And Seth lived an hundred and five years , and bega
ve years , and begat Enos : And Seth lived after he begat Enos eight hundred an
welve years : and he died . And Enos lived ninety years , and begat Cainan : An
 years , and begat Cainan : And Enos lived after he begat Cainan eight hundred 
ive years : and he died . And Cainan lived seventy years and begat Mahalaleel :
rs and begat Mahalaleel : And Cainan lived after he begat Mahalaleel eight hund
years : and he died . And Mahalaleel lived sixty and five years , and begat Jar
s , and begat Jared : And Mahalaleel lived after he begat Jared eight hundred a
and five yea and he died . And Jared lived an hundred sixty and two years , and
o years , and he begat Eno And Jared lived after he begat Enoch eight hundred y
 and two ye

A concordance permits us to see words in context. For example, we saw that monstrous occurred in contexts such as the ___ pictures and a ___ size . What other words appear in a similar range of contexts? We can find out by appending the term similar to the name of the text in question, then inserting the relevant word in parentheses:

In [8]:
text1.similar("monstrous")

true contemptible christian abundant few part mean careful puzzled
mystifying passing curious loving wise doleful gamesome singular
delightfully perilous fearless


In [9]:
text2.similar("monstrous")

very so exceedingly heartily a as good great extremely remarkably
sweet vast amazingly


We got different results for different texts. Austen uses this word quite differently from Melville; for her, monstrous has positive connotations, and sometimes functions as an intensifier like the word very.

The term common_contexts allows us to examine just the contexts that are shared by two or more words, such as monstrous and very.

In [10]:
text2.common_contexts(["monstrous", "very"])

a_pretty am_glad a_lucky is_pretty be_glad


In [11]:
text2.common_contexts(["monstrous", "sweet"])

a_pretty


In [12]:
text1.common_contexts(["monstrous", "true"])

the_pictures


We can also determine the location of a word in the text: how many words from the beginning it appears. This positional information can be displayed using a dispersion plot.

In [13]:
text4.dispersion_plot(["citizens", "democracy", "freedom", "duties", "America"])

<Figure size 640x480 with 1 Axes>

Now, just for fun, let's try generating some random text in the various styles we have just seen. To do this, we type the name of the text followed by the term generate. (We need to include the parentheses, but there's nothing that goes between them.)

In [14]:
text3.generate()

Building ngram index...


laid by her , and said unto Cain , Where art thou , and said , Go to ,
I will not do it for ten ' s sons ; we dreamed each man according to
their generatio the firstborn said unto Laban , Because I said , Nay ,
but Sarah shall her name be . , duke Elah , duke Shobal , and Akan .
and looked upon my affliction . Bashemath Ishmael ' s blood , but Isra
for as a prince hast thou found of all the cattle in the valley , and
the wo The


"laid by her , and said unto Cain , Where art thou , and said , Go to ,\nI will not do it for ten ' s sons ; we dreamed each man according to\ntheir generatio the firstborn said unto Laban , Because I said , Nay ,\nbut Sarah shall her name be . , duke Elah , duke Shobal , and Akan .\nand looked upon my affliction . Bashemath Ishmael ' s blood , but Isra\nfor as a prince hast thou found of all the cattle in the valley , and\nthe wo The"

### Counting Vocabulary

Let's begin by finding out the length of a text from start to finish, in terms of the words and punctuation symbols that appear. We use the term len to get the length of something, which we'll apply here to the book of Genesis:

In [15]:
len(text3)

44764

The vocabulary of a text is just the set of tokens that it uses, since in a set, all duplicates are collapsed together. In Python we can obtain the vocabulary items of text3 with the command: set(text3).

In [16]:
sorted(set(text3))

['!',
 "'",
 '(',
 ')',
 ',',
 ',)',
 '.',
 '.)',
 ':',
 ';',
 ';)',
 '?',
 '?)',
 'A',
 'Abel',
 'Abelmizraim',
 'Abidah',
 'Abide',
 'Abimael',
 'Abimelech',
 'Abr',
 'Abrah',
 'Abraham',
 'Abram',
 'Accad',
 'Achbor',
 'Adah',
 'Adam',
 'Adbeel',
 'Admah',
 'Adullamite',
 'After',
 'Aholibamah',
 'Ahuzzath',
 'Ajah',
 'Akan',
 'All',
 'Allonbachuth',
 'Almighty',
 'Almodad',
 'Also',
 'Alvah',
 'Alvan',
 'Am',
 'Amal',
 'Amalek',
 'Amalekites',
 'Ammon',
 'Amorite',
 'Amorites',
 'Amraphel',
 'An',
 'Anah',
 'Anamim',
 'And',
 'Aner',
 'Angel',
 'Appoint',
 'Aram',
 'Aran',
 'Ararat',
 'Arbah',
 'Ard',
 'Are',
 'Areli',
 'Arioch',
 'Arise',
 'Arkite',
 'Arodi',
 'Arphaxad',
 'Art',
 'Arvadite',
 'As',
 'Asenath',
 'Ashbel',
 'Asher',
 'Ashkenaz',
 'Ashteroth',
 'Ask',
 'Asshur',
 'Asshurim',
 'Assyr',
 'Assyria',
 'At',
 'Atad',
 'Avith',
 'Baalhanan',
 'Babel',
 'Bashemath',
 'Be',
 'Because',
 'Becher',
 'Bedad',
 'Beeri',
 'Beerlahairoi',
 'Beersheba',
 'Behold',
 'Bela',
 'Belah

In [17]:
len(set(text3))

2789

By wrapping sorted() around the Python expression set(text3), we obtain a sorted list of vocabulary items, beginning with various punctuation symbols and continuing with words starting with A. All capitalized words precede lowercase words. We discover the size of the vocabulary indirectly, by asking for the number of items in the set, and again we can use len to obtain this number. Although it has 44,764 tokens, this book has only 2,789 distinct words, or "word types." A word type is the form or spelling of the word independently of its specific occurrences in a text — that is, the word considered as a unique item of vocabulary. Our count of 2,789 items will include punctuation symbols, so we will generally call these unique items types instead of word types.

Now, let's calculate a measure of the lexical richness of the text. The next example shows us that the number of distinct words is just 6% of the total number of words, or equivalently that each word is used 16 times on average (remember if you're using Python 2, to start with from __future__ import division).

In [18]:
len(set(text3)) / len(text3)

0.06230453042623537

Next, let's focus on particular words. We can count how often a word occurs in a text, and compute what percentage of the text is taken up by a specific word:

In [19]:
text3.count("smote")

5

In [20]:
100 * text4.count('a') / len(text4)

1.457973123627309

In [21]:
text3.count("lol")

0

In [22]:
text5.count("lol")

704

In [23]:
100 * text5.count('lol') / len(text5)

1.5640968673628082

We found that text3: The Book of Genesis does not contain word "lol" at all. text5: Chat Corpus has 704 of words "lol", which is 1.56% of the whole text. 

The next example shows how to define two new functions, lexical_diversity() and  percentage():

In [24]:
def lexical_diversity(text):
     return len(set(text)) / len(text)
    
def percentage(count, total):
     return 100 * count / total    

In [25]:
lexical_diversity(text3)

0.06230453042623537

In [26]:
lexical_diversity(text5)

0.13477005109975562

In [27]:
percentage(4, 5)

80.0

In [28]:
percentage(text4.count('a'), len(text4))

1.457973123627309

### 3. Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

In [36]:
f1=open("HW/books/14640.txt")
line1 =f1.readlines()[0:2]
print (line1)


["The Project Gutenberg EBook of McGuffey's First Eclectic Reader, Revised\n", 'Edition, by William Holmes McGuffey\n']


In [37]:
f2=open("HW/books/14766.txt", "r")
book2 =f2.readlines()[0:2]
print (book2)

["The Project Gutenberg EBook of McGuffey's Third Eclectic Reader\n", 'by William Holmes McGuffey\n']


In [38]:
f3=open("HW/books/16751.txt", "r")
book3 =f3.readlines()[0:2]
print (book3)

["The Project Gutenberg EBook of McGuffey's Sixth Eclectic Reader\n", 'by William Holmes McGuffey\n']


In [43]:
f1=open("HW/books/14640.txt")
book1 =f1.read()
tokens1 = nltk.word_tokenize(book1)
txt1 = nltk.Text(tokens1)
lexical_diversity1 = lexical_diversity(txt1)
vocab1 = len(set(txt1))
len1 = len(txt1)
name1 = 'The Project Gutenberg EBook of McGuffeys First Eclectic Reader'

In [44]:
f2=open("HW/books/14766.txt", "r")
book2 =f2.read()
tokens2 = nltk.word_tokenize(book2)
txt2 = nltk.Text(tokens2)
lexical_diversity2 = lexical_diversity(txt2)
vocab2 = len(set(txt2))
len2 = len(txt2)
name2 = 'The Project Gutenberg EBook of McGuffeys Third Eclectic Reader'

In [46]:
f3=open("HW/books/16751.txt", "r")
book3 =f3.read()
tokens3 = nltk.word_tokenize(book3)
txt3 = nltk.Text(tokens3)
lexical_diversity3 = lexical_diversity(txt3)
vocab3 = len(set(txt3))
len3 = len(txt3)
name3 = 'The Project Gutenberg EBook of McGuffeys Sixth Eclectic Reader'

In [47]:
from tabulate import tabulate
print(tabulate([[name1, lexical_diversity1,vocab1, len1], [name2, lexical_diversity2,vocab2,len2], [name3, lexical_diversity3,vocab3,len3]], headers=['Book','Lexical diversity', 'Vocabulary size', 'Text length']))

Book                                                              Lexical diversity    Vocabulary size    Text length
--------------------------------------------------------------  -------------------  -----------------  -------------
The Project Gutenberg EBook of McGuffeys First Eclectic Reader             0.167108               2147          12848
The Project Gutenberg EBook of McGuffeys Third Eclectic Reader             0.124148               4715          37979
The Project Gutenberg EBook of McGuffeys Sixth Eclectic Reader             0.10088               17260         171094


### 4. Also compare the vocabulary size of the same three texts. Explain whether the result was surprising. 
Vocabulary size for small kids is 2147, and older readers is 17260. This looks logical, sinse older readers should have stronger vocabulary than smaller kids.

### 5. Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.
Vocabulary size and lexical diversity in combination will be a better measure of text difficulty (or reading level). Sinse lexical diversity and vocabulary size are sensetive to the text length, it is better to use them as a measure of text difficulty in combination, sinse by itself they do not represent the real picture like we saw in the example with The Project Gutenberg EBook of McGuffeys. With increasing of text length you will get decreasing of Lexical diversity and the main reason, I think, may be function words.